### PR TITLE
Add hstackBlocks for Block to join blocks by columns

### LINKED ISSUE #5829: Support late materialization
## Feature Request

**Is your feature request related to a problem? Please describe:**

For some query like `select * from t where a + b <> 0;`, we now scan all columns and then do filter, which is heavy and we can do better.

**Describe the feature you'd like:**

Support late materialization. We can scan column(s) at first to filter out unnecessary data segment, and then just scan the rest columns.

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
Signed-off-by: Lloyd-Pottiger <yan1579196623@gmail.com>

### What problem does this PR solve?

Issue Number: close #5829 

Problem Summary:

### What is changed and how it works?

1. rename `mergeBlocks` to `vstackBlocks`, join blocks by rows
2. add `hstackBlocks`, join blocks by columns

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Core/Block.cpp [hunk 1]
```diff
@@ -526,33 +548,33 @@ Block mergeBlocks(Blocks && blocks)
...
+    Block out = blocks.front().cloneEmpty();
+    MutableColumns columns = out.mutateColumns();
+    for (auto & col : columns)
     {
-        dst_columns[i] = (*std::move(first_block.getByPosition(i).column)).mutate();
-        dst_columns[i]->reserve(result_rows);
+        col->reserve(result_rows);
     }
 
     for (size_t i = 1; i < blocks.size(); ++i)
```
Comment (SeaRise) [inline dbms/src/Core/Block.cpp:564]: Is the first block ignored?

Comment (Lloyd-Pottiger) [reply dbms/src/Core/Block.cpp]: sorry, bug, fixed.

File: dbms/src/Core/Block.cpp [hunk 2]
```diff
@@ -514,7 +514,29 @@ static ReturnType checkBlockStructure(const Block & lhs, const Block & rhs, cons
     return ReturnType(true);
 }
 
-Block mergeBlocks(Blocks && blocks)
+/// join blocks by columns
+Block hstackBlocks(Blocks && blocks, const Block & header)
```
Comment (SeaRise) [inline dbms/src/Core/Block.cpp:518]: Is `header` needed here?
Can we use `blocks.front().cloneEmpty()`?

File: dbms/src/Core/Block.cpp [hunk 3]
```diff
@@ -514,7 +514,29 @@ static ReturnType checkBlockStructure(const Block & lhs, const Block & rhs, cons
...
+
+    Block res = header.cloneEmpty();
+
+    size_t num_rows = blocks.front().rows();
+    for (const auto & block : blocks)
+    {
+        RUNTIME_CHECK_MSG(block.rows() == num_rows, "Cannot hstack blocks with different number of rows");
+        for (const auto & elem : block)
+        {
+            res.getByName(elem.name).column = std::move(elem.column);
```
Comment (SeaRise) [inline dbms/src/Core/Block.cpp:531]: Maybe we should check `elem.name` unique here.

**Review by SeaRise — COMMENTED** [review]
Others LGTM

File: dbms/src/Core/Block.cpp [hunk 4]
```diff
@@ -514,7 +514,29 @@ static ReturnType checkBlockStructure(const Block & lhs, const Block & rhs, cons
...
+        }
+    }
+
+    return res;
+}
+
+/// join blocks by rows
+Block vstackBlocks(Blocks && blocks)
```
Comment (SeaRise) [inline dbms/src/Core/Block.cpp:539]: The name confuses me a bit...

**Review by SeaRise — APPROVED** [review]

**Review by ywqzzy — APPROVED** [review]

File: dbms/src/Core/Block.cpp (hunk 4 above)
Comment (Lloyd-Pottiger) [reply dbms/src/Core/Block.cpp:539]: yeah, any better idea?

Comment (SeaRise) [reply dbms/src/Core/Block.cpp:539]: I have no idea about it,

**Pull request merged**